From 11af27b6f4438975a090dd6bdac15f3b007cd0f2 Mon Sep 17 00:00:00 2001
From: Dmitry Artamonow <mad_soft@inbox.ru>
Date: Sat, 21 Mar 2009 02:05:34 +0300
Subject: [PATCH 28/28] h3600: fix suspend (HACK)

Due to bug in Compaq bootldr suspend doesn't work since kernel 2.6.27
(to be more precise since commit 5f2809e69c7128f86316048221cf45146f69a4a0)
Here we're trying to fool bootldr by changing values in checked memory
area before suspending, and restoring them after wakeup. As a side effect,
this will probably fix suspend on kernels booted by HaRET too.

Signed-off-by: Dmitry Artamonow <mad_soft@inbox.ru>
---
 arch/arm/mach-sa1100/pm.c |   18 ++++++++++++++++++
 1 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-sa1100/pm.c b/arch/arm/mach-sa1100/pm.c
index 111cce6..24db59a 100644
--- a/arch/arm/mach-sa1100/pm.c
+++ b/arch/arm/mach-sa1100/pm.c
@@ -56,6 +56,21 @@ static int sa11x0_pm_enter(suspend_state_t state)
 {
 	unsigned long gpio, sleep_save[SLEEP_SAVE_COUNT];
 
+#if defined(CONFIG_SA1100_H3100) || defined(CONFIG_SA1100_H3600)
+/* Compaq bootldr is trying to detect is Linux or WinCE needs resuming,	*
+ * but detection is poor and works incorrectly with kernels >=2.6.27	*
+ * Since bootldr development is dead, we trying to workaround this in	*
+ * kernel instead. Bootldr tests some bytes at 0xC0001000 and if they	*
+ * are equal to zero, it tries to wake Windows (even if it's erased	*
+ * from flash). So we are fooling bootldr by putting some junk at this	*
+ * testpoint address							*/
+	unsigned long *bootldr_testpoint  = phys_to_virt(PHYS_OFFSET + SZ_4K);
+	unsigned long saved_testpoint;
+
+	saved_testpoint = *bootldr_testpoint;
+	*bootldr_testpoint = 0xdeadbeef;
+#endif
+
 	gpio = GPLR;
 
 	/* save vital registers */
@@ -107,6 +122,9 @@ static int sa11x0_pm_enter(suspend_state_t state)
 	GPSR = gpio;
 	GPCR = ~gpio;
 
+#if defined(CONFIG_SA1100_H3100) || defined(CONFIG_SA1100_H3600)
+	*bootldr_testpoint = saved_testpoint;
+#endif
 	/*
 	 * Clear the peripheral sleep-hold bit.
 	 */
-- 
1.6.1.3

